app.go

```diff
@@ -35,8 +35,19 @@ import (
 	"go.uber.org/multierr"
 )
 
+type appState int
```

Comment:
Let's jam in person. 

I think that `fx.App` might actually want a state machine repr like this so that we can help people interact with it correctly. If we can come to the same thinking here, then I'd probably want to see this commit  broken up into a stack:

1. Add `fx.App.state` with validations
2. Add `fx.App.Invoke` to invoke a func while running